src/systems/equation_systems.C

```diff
@@ -146,28 +146,10 @@ bool EquationSystems::reinit_solutions ()
   const unsigned int n_sys = this->n_systems();
   libmesh_assert_not_equal_to (n_sys, 0);
 
-  // We may have added new systems since our last
-  // EquationSystems::(re)init call
-  bool _added_new_systems = false;
+  // And any new systems will need initialization
   for (unsigned int i=0; i != n_sys; ++i)
     if (!this->get_system(i).is_initialized())
-      _added_new_systems = true;
-
-  if (_added_new_systems)
-    {
-      // Our DofObjects will need space for the additional systems
-      for (auto & node : _mesh.node_ptr_range())
-        node->set_n_systems(n_sys);
-
-      for (auto & elem : _mesh.element_ptr_range())
-        elem->set_n_systems(n_sys);
-
-      // And any new systems will need initialization
-      for (unsigned int i=0; i != n_sys; ++i)
-        if (!this->get_system(i).is_initialized())
-          this->get_system(i).init();
-    }
-
+      this->get_system(i).init();
```

Comment:
I'll let @roystgnr comment here, but it looks to me like you are removing some `set_n_systems()` calls entirely. I guess what you are saying is that this happens somewhere else instead, so this code is redundant?

Reply:
Looks redundant to me, actually.  We do that later sweep over `set_n_systems()` regardless, to handle the "users added nodes or elements after system initialization", and I guess whoever added that didn't realize that it made the `set_n_systems()` calls from the "new system has been added since last init" case superfluous.

The most efficient thing to do would be to ensure that new nodes and elements get `set_n_systems()` done as they're added, though.  We should probably add a `MeshBase::set_n_systems()` method, which would first do the loops below but would then cache the `n_sys` value and use it on new additions too, and then we'd *really* only need to do a full loop when new post-initialization Systems get added.

Reply:
(That's an idea for another PR, though; this is still a big improvement even if it could be made better still)